### PR TITLE
[`flake8-pyi`] Implement `PYI042` and `PYI043`

### PR DESCRIPTION
**Pull request by arya-k**
PYI042 Type alias names should use CamelCase rather than snake_case
PYI043 Do not use names ending in "T" for private type aliases. (The "T" suffix implies that an object is a TypeVar.)

rel: https://github.com/charliermarsh/ruff/issues/848

### CODE REVIEW

**Comment by arya-k** [comment]
Hmm, I'm still pretty new to the project and not quite sure why this is failing– I can't repro locally with `cargo test` (macOS x86), and these integration tests don't seem related to my code. Do you have any suggestions @charliermarsh?

(also- somewhat new to this project. Is it bad form to tag you like this?)

**Comment by arya-k** [comment]
Hmm I'm not able to repro this test failure locally (macOS), when I run `cargo test -p ruff_cli --test integration_test`. I'm not quite sure what to make of this, and the CLI tests seem unrelated to my changes– do you have any tips @charliermarsh?

p.s. I'm pretty new to the ruff community: is it bad form to tag you like this?

**Comment by charliermarsh** [comment]
@arya-k - Not sure why I can't repro locally either, but I have a suspicion that we just hit a factor-of-two boundary and need to increase the `RuleSet` bit vector size. In short, we have a vector that needs to be long enough to match the number of total rules, and we may have just exceeded the previous limit. I tried pushing a fix, lets see what happens.

Also: totally fine to tag me or other maintainers.

File: crates/ruff/src/registry/rule_set.rs [hunk 1]
```diff
@@ -7,10 +7,10 @@ use std::iter::FusedIterator;
 ///
 /// Uses a bitset where a bit of one signals that the Rule with that [u16] is in this set.
 #[derive(Clone, Default, CacheKey, PartialEq, Eq)]
-pub struct RuleSet([u64; 9]);
+pub struct RuleSet([u64; 10]);
 
 impl RuleSet {
-    const EMPTY: [u64; 9] = [0; 9];
+    const EMPTY: [u64; 10] = [0; 10];
```
Comment (charliermarsh) [inline crates/ruff/src/registry/rule_set.rs:13]: @arya-k - Here's the change I was referring to (very unclear from the trace, but I saw the out-of-bounds with index 9 and thought of the 9 here).

Comment (arya-k) [reply crates/ruff/src/registry/rule_set.rs:13]: Thank you! This is helpful, and I appreciate the fix. 

I wonder why this didn’t come up locally? I’ll do some investigating later.

**Review by MichaReiser — COMMENTED** [review]
Thank you for your contribution. 

I think we should try to implement these rules without having to rely on regular expressions. Regular expressions are great but can be overkill when matching on simple patterns and using plain string methods can then be better for performance

File: crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
+use once_cell::sync::Lazy;
+use regex::Regex;
+use rustpython_parser::ast::{Expr, ExprKind};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+
+use crate::checkers::ast::Checker;
+
+static CAMEL_CASE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"^_?[a-z]").unwrap());
```
[the remark this answers is not in the document]
Comment (arya-k) [reply crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs:10]: We should match that pattern! I also realize that the regex should have been named `SNAKE_CASE_REGEX`, which may have also added to the confusion. I'll switch over to the new format!

File: crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        if let Some(c) = chars.next() {
+            first = c;
+        } else {
+            return false;
+        }
+    }
+
+    // There should be at least one other underscore in the name.
+    first.is_ascii_alphabetic() && chars.any(|c| c == '_')
+}
```
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs:63]: I did some testing against `typeshed`, and I think this matches their current behavior:

```rust
/// Return `true` if the given name is a snake_case type alias. In this context, we match against
/// any name that begins with an optional underscore, followed by at least one lowercase letter.
fn is_snake_case_type_alias(name: &str) -> bool {
    let mut chars = name.chars();
    match (chars.next(), chars.next()) {
        (Some('_'), Some(c)) | (Some(c), ..) => c.is_ascii_alphabetic() && c.is_ascii_lowercase(),
        _ => false,
    }
}
```

Comment (charliermarsh) [reply crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs:63]: The current version seems to have some false positives, e.g., it flags this, which `flake8-pyi` allows:

```py
_TYPE_FIELDS: TypeAlias = _TYPE_FIELDS_SEQUENCE | Mapping[str, fields._FieldValueTuple]
```

Comment (arya-k) [reply crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs]: That makes sense- and is the original behavior the regex supported (`^_?[a-z]`).

I changed it to support `Snake_case`- is it more important to maintain parity with the original plug-in?

**Comment by arya-k** [comment]
Thank you for pushing this over the finish line- for my understanding, how did you format the python files? My installation of `black` produced a much larger diff so I ignored it.

File: crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs]: Ah yeah, I prefer to maintain parity with upstream plugins unless we have a good reason to deviate (examples: there's a clear bug that we can fix, diagnostic messages are confusing or inconsistent with our own conventions). It makes migrations easier for users, and reduces confusion.

**Comment by charliermarsh** [comment]
@arya-k - Oh interesting, I just ran Black from my own editor.

**Comment by charliermarsh** [comment]
I should correct that: we probably haven't run Black over all of our Python fixtures, so running `black` from root could produce a large diff. I just ran Black over the new fixtures in this PR.

**Comment by arya-k** [comment]
> I should correct that: we probably haven't run Black over all of our Python fixtures, so running `black` from root could produce a large diff. I just ran Black over the new fixtures in this PR.

Ah got it. I ran it on all fixtures, then figured it wasn’t the way we were formatting python. Thanks for the clarification!

**Pull request merged**